pkg/apiserver/user/sqlauth/sqlauth.go

```diff
@@ -60,3 +98,81 @@ func (a *Authenticator) Authenticate(f user.AuthenticateForm) (*utils.SessionUse
 		IsWriteable:  true,
 	}, nil
 }
+
+// Currently, There are 2 kinds of grant output format in TiDB:
+// - GRANT [grants] ON [db.table] TO [user]
+// - GRANT [roles] TO [user]
+// Examples:
+// - GRANT PROCESS,SHOW DATABASES,CONFIG ON *.* TO 'dashboardAdmin'@'%'
+// - GRANT SYSTEM_VARIABLES_ADMIN,RESTRICTED_VARIABLES_ADMIN,RESTRICTED_STATUS_ADMIN,RESTRICTED_TABLES_ADMIN ON *.* TO 'dashboardAdmin'@'%'
+// - GRANT ALL PRIVILEGES ON *.* TO 'dashboardAdmin'@'%'
+// - GRANT `app_read`@`%` TO `test`@`%`
+func parseCurUserGrants(grantRows []string) []string {
+	grants := make([]string, 0)
+	grantRegex := regexp.MustCompile(`GRANT (.+) ON`)
+
+	for _, row := range grantRows {
+		m := grantRegex.FindStringSubmatch(row)
+		if len(m) == 2 {
+			curRowGrants := strings.Split(m[1], ",")
+			grants = append(grants, curRowGrants...)
+		}
+	}
+
+	return grants
+}
+
+// Following base privileges are required
+// - ALL PRIVILEGES
+// - or
+// - PROCESS
+// - SHOW DATABASES
+// - CONFIG
+// - SYSTEM_VARIABLES_ADMIN or SUPER (SUPER includes SYSTEM_VARIABLES_ADMIN)
+// - DASHBOARD_CLIENT or SUPER (SUPER includes DASHBOARD_CLIENT)
+// When TiDB SEM is enabled, following extra privileges are required
+// - RESTRICTED_VARIABLES_ADMIN
+// - RESTRICTED_TABLES_ADMIN
+// - RESTRICTED_STATUS_ADMIN
+func checkDashboardPrivileges(grants []string, enableSEM bool) bool {
+	hasPriv := func(priv string) bool {
+		for _, grant := range grants {
+			if priv == grant {
+				return true
+			}
+		}
+		return false
+	}
+
+	if !hasPriv("ALL PRIVILEGES") {
+		if !hasPriv("PROCESS") {
+			return false
+		}
+		if !hasPriv("SHOW DATABASES") {
+			return false
+		}
+		if !hasPriv("CONFIG") {
+			return false
+		}
+		if !hasPriv("SUPER") {
+			if !hasPriv("SYSTEM_VARIABLES_ADMIN") {
+				return false
+			}
+			if !hasPriv("DASHBOARD_CLIENT") {
+				return false
+			}
+		}
+	}
+	if enableSEM {
```

Comment:
Does `ALL PRIVILEGES` contain the following `RESTRICTED_` privileges? @crazycs520

Reply:
No, `ALL PRIVILEGES` doesn't contain the `RESTRICTED_xxx` privileges.